File: hello-java/audio-stream-client/src/main/java/org/feuyeux/mmap/audio/client/util/StreamIdGenerator.java
package org.feuyeux.mmap.audio.client.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.UUID;
import java.util.regex.Pattern;

/**
 * Stream ID generator for creating unique stream identifiers.
 * Matches the C++ StreamIdGenerator interface.
 */
public class StreamIdGenerator {
    private static final Logger logger = LoggerFactory.getLogger(StreamIdGenerator.class);
    private static final String DEFAULT_PREFIX = "stream";
    private static final Pattern STREAM_ID_PATTERN = Pattern.compile("^[a-zA-Z0-9_-]+-[a-f0-9]{8}(-[a-f0-9]{4}){3}-[a-f0-9]{12}$");

    /**
     * Generate a unique stream ID with default prefix "stream".
     *
     * @return stream ID in format "stream-{uuid}"
     */
    public String generate() {
        return generateWithPrefix(DEFAULT_PREFIX);
    }

    /**
     * Generate a unique stream ID with custom prefix.
     *
     * @param prefix prefix for the stream ID
     * @return stream ID in format "{prefix}-{uuid}"
     */
    public String generateWithPrefix(String prefix) {
        if (prefix == null || prefix.isEmpty()) {
            prefix = DEFAULT_PREFIX;
        }
        
        String uuid = UUID.randomUUID().toString();
        String streamId = prefix + "-" + uuid;
        
        logger.debug("Generated stream ID: {}", streamId);
        return streamId;
    }

    /**
     * Generate a short stream ID (8 characters).
     *
     * @return short stream ID in format "stream-{short-uuid}"
     */
    public String generateShort() {
        return generateShortWithPrefix(DEFAULT_PREFIX);
    }

    /**
     * Generate a short stream ID with custom prefix.
     *
     * @param prefix prefix for the stream ID
     * @return short stream ID in format "{prefix}-{short-uuid}"
     */
    public String generateShortWithPrefix(String prefix) {
        if (prefix == null || prefix.isEmpty()) {
            prefix = DEFAULT_PREFIX;
        }
        
        String uuid = UUID.randomUUID().toString().substring(0, 8);
        String streamId = prefix + "-" + uuid;
        
        logger.debug("Generated short stream ID: {}", streamId);
        return streamId;
    }

    /**
     * Validate a stream ID format.
     *
     * @param streamId stream ID to validate
     * @return true if valid format
     */
    public boolean validate(String streamId) {
        if (streamId == null || streamId.isEmpty()) {
            logger.warn("Stream ID is null or empty");
            return false;
        }
        
        // Check if it matches the expected pattern
        boolean isValid = STREAM_ID_PATTERN.matcher(streamId).matches();
        
        if (!isValid) {
            logger.warn("Invalid stream ID format: {}", streamId);
        }
        
        return isValid;
    }

    /**
     * Validate a short stream ID format.
     *
     * @param streamId stream ID to validate
     * @return true if valid short format
     */
    public boolean validateShort(String streamId) {
        if (streamId == null || streamId.isEmpty()) {
            logger.warn("Stream ID is null or empty");
            return false;
        }
        
        // Check if it matches the short pattern: prefix-8chars
        Pattern shortPattern = Pattern.compile("^[a-zA-Z0-9_-]+-[a-f0-9]{8}$");
        boolean isValid = shortPattern.matcher(streamId).matches();
        
        if (!isValid) {
            logger.warn("Invalid short stream ID format: {}", streamId);
        }
        
        return isValid;
    }

    /**
     * Extract the prefix from a stream ID.
     *
     * @param streamId stream ID
     * @return prefix, or null if invalid format
     */
    public String extractPrefix(String streamId) {
        if (streamId == null || streamId.isEmpty()) {
            return null;
        }
        
        int dashIndex = streamId.indexOf('-');
        if (dashIndex > 0) {
            return streamId.substring(0, dashIndex);
        }
        
        return null;
    }

    /**
     * Extract the UUID part from a stream ID.
     *
     * @param streamId stream ID
     * @return UUID string, or null if invalid format
     */
    public String extractUuid(String streamId) {
        if (streamId == null || streamId.isEmpty()) {
            return null;
        }
        
        int dashIndex = streamId.indexOf('-');
        if (dashIndex > 0 && dashIndex < streamId.length() - 1) {
            return streamId.substring(dashIndex + 1);
        }
        
        return null;
    }
}

File: hello-csharp/src/Client/Util/StreamIdGenerator.cs
using System;
using System.Text.RegularExpressions;

namespace AudioFileTransfer.Client.Util;

/// <summary>
/// Stream ID generator for creating unique stream identifiers
/// Matches the Java StreamIdGenerator interface.
/// </summary>
public static class StreamIdGenerator
{
    private const string DEFAULT_PREFIX = "stream";

    /// <summary>
    /// Generate a unique stream ID with default prefix "stream".
    /// </summary>
    /// <returns>Stream ID in format "stream-{uuid}"</returns>
    public static string Generate()
    {
        return GenerateWithPrefix(DEFAULT_PREFIX);
    }

    /// <summary>
    /// Generate a unique stream ID with custom prefix.
    /// </summary>
    /// <param name="prefix">Prefix for the stream ID</param>
    /// <returns>Stream ID in format "{prefix}-{uuid}"</returns>
    public static string GenerateWithPrefix(string prefix)
    {
        if (string.IsNullOrEmpty(prefix))
        {
            prefix = DEFAULT_PREFIX;
        }
        
        string uuid = Guid.NewGuid().ToString();
        string streamId = $"{prefix}-{uuid}";
        
        return streamId;
    }

    /// <summary>
    /// Generate a short stream ID (8 characters).
    /// </summary>
    /// <returns>Short stream ID in format "stream-{short-uuid}"</returns>
    public static string GenerateShort()
    {
        return GenerateShortWithPrefix(DEFAULT_PREFIX);
    }

    /// <summary>
    /// Generate a short stream ID with custom prefix.
    /// </summary>
    /// <param name="prefix">Prefix for the stream ID</param>
    /// <returns>Short stream ID in format "{prefix}-{short-uuid}"</returns>
    public static string GenerateShortWithPrefix(string prefix)
    {
        if (string.IsNullOrEmpty(prefix))
        {
            prefix = DEFAULT_PREFIX;
        }
        
        string uuid = Guid.NewGuid().ToString("N").Substring(0, 8);
        string streamId = $"{prefix}-{uuid}";
        
        return streamId;
    }

    /// <summary>
    /// Validate a stream ID format.
    /// </summary>
    /// <param name="streamId">Stream ID to validate</param>
    /// <returns>True if valid format</returns>
    public static bool Validate(string streamId)
    {
        if (string.IsNullOrEmpty(streamId))
        {
            return false;
        }
        
        // Check if it matches the expected pattern
        Regex pattern = new Regex(@"^[a-zA-Z0-9_-]+-[a-f0-9]{8}(-[a-f0-9]{4}){3}-[a-f0-9]{12}$");
        return pattern.IsMatch(streamId);
    }

    /// <summary>
    /// Validate a short stream ID format.
    /// </summary>
    /// <param name="streamId">Stream ID to validate</param>
    /// <returns>True if valid short format</returns>
    public static bool ValidateShort(string streamId)
    {
        if (string.IsNullOrEmpty(streamId))
        {
            return false;
        }
        
        // Check if it matches the short pattern: prefix-8chars
        Regex pattern = new Regex(@"^[a-zA-Z0-9_-]+-[a-f0-9]{8}$");
        return pattern.IsMatch(streamId);
    }

    /// <summary>
    /// Extract the prefix from a stream ID.
    /// </summary>
    /// <param name="streamId">Stream ID</param>
    /// <returns>Prefix, or null if invalid format</returns>
    public static string? ExtractPrefix(string streamId)
    {
        if (string.IsNullOrEmpty(streamId))
        {
            return null;
        }
        
        int dashIndex = streamId.IndexOf('-');
        if (dashIndex > 0)
        {
            return streamId.Substring(0, dashIndex);
        }
        
        return null;
    }

    /// <summary>
    /// Extract the UUID part from a stream ID.
    /// </summary>
    /// <param name="streamId">Stream ID</param>
    /// <returns>UUID string, or null if invalid format</returns>
    public static string? ExtractUuid(string streamId)
    {
        if (string.IsNullOrEmpty(streamId))
        {
            return null;
        }
        
        int dashIndex = streamId.IndexOf('-');
        if (dashIndex > 0 && dashIndex < streamId.Length - 1)
        {
            return streamId.Substring(dashIndex + 1);
        }
        
        return null;
    }
}
